File: lib/views/map/weapon.dart
class Weapon {
  final String name;
  final double range; // in meters
  final double mil; // in mils
  final double flightTime; // in seconds
  final int gunPower; // arbitrary unit
  final int id; // ID for each weapon

  Weapon({
    required this.name,
    required this.range,
    required this.mil,
    required this.flightTime,
    required this.gunPower,
    required this.id,
  });
}

List<Weapon> generateWeapons() {
  List<Weapon> weapons = [];
  int? id = 1;

  // Add MA7 data
  const List<Map<String, dynamic>> ma7Data = [
    {
      "gunPower": 0,
      "minRange": 250,
      "maxRange": 750,
      "minMil": 1427,
      "maxMil": 800,
      "minTime": 18.3,
      "maxTime": 13.1
    },
    {
      "gunPower": 1,
      "minRange": 500,
      "maxRange": 1550,
      "minMil": 1437,
      "maxMil": 800,
      "minTime": 25.3,
      "maxTime": 18.2
    },
    {
      "gunPower": 2,
      "minRange": 800,
      "maxRange": 2150,
      "minMil": 1408,
      "maxMil": 952,
      "minTime": 25.9,
      "maxTime": 18.6
    },
    {
      "gunPower": 3,
      "minRange": 1000,
      "maxRange": 2900,
      "minMil": 1421,
      "maxMil": 800,
      "minTime": 40.2,
      "maxTime": 28.8
    },
    {
      "gunPower": 4,
      "minRange": 1300,
      "maxRange": 3480,
      "minMil": 1405,
      "maxMil": 800,
      "minTime": 46.1,
      "maxTime": 33.1
    },
    {
      "gunPower": 5,
      "minRange": 1600,
      "maxRange": 4050,
      "minMil": 1394,
      "maxMil": 800,
      "minTime": 51.5,
      "maxTime": 37.2
    },
    {
      "gunPower": 6,
      "minRange": 1800,
      "maxRange": 4600,
      "minMil": 1395,
      "maxMil": 800,
      "minTime": 57.0,
      "maxTime": 41.0
    },
  ];
  weapons.addAll(generateIntervalData("MA7", ma7Data, id));
  id += ((ma7Data.length *
      ((ma7Data[0]["maxRange"] - ma7Data[0]["minRange"]) ~/ 50)) as int?)!;

  List<Map<String, dynamic>> ma8Data = [
    {
      "gunPower": 0,
      "minRange": 200,
      "maxRange": 550,
      "minMil": 1410.0,
      "maxMil": 800.0,
      "minTime": 14.0,
      "maxTime": 10.0
    },
    {
      "gunPower": 1,
      "minRange": 400,
      "maxRange": 1125,
      "minMil": 1414.0,
      "maxMil": 800.0,
      "minTime": 21.0,
      "maxTime": 15.1
    },
    {
      "gunPower": 2,
      "minRange": 600,
      "maxRange": 1650,
      "minMil": 1410.0,
      "maxMil": 800.0,
      "minTime": 27.2,
      "maxTime": 17.6
    },
    {
      "gunPower": 3,
      "minRange": 900,
      "maxRange": 2700,
      "minMil": 1427.0,
      "maxMil": 800.0,
      "minTime": 36.4,
      "maxTime": 26.1
    },
    {
      "gunPower": 4,
      "minRange": 1300,
      "maxRange": 3550,
      "minMil": 1409.0,
      "maxMil": 800.0,
      "minTime": 44.2,
      "maxTime": 45.0
    },
    {
      "gunPower": 5,
      "minRange": 1600,
      "maxRange": 4400,
      "minMil": 1412.0,
      "maxMil": 868.0,
      "minTime": 51.1,
      "maxTime": 39.1
    },
    {
      "gunPower": 6,
      "minRange": 2100,
      "maxRange": 5150,
      "minMil": 1387.0,
      "maxMil": 800.0,
      "minTime": 57.8,
      "maxTime": 41.8
    },
    {
      "gunPower": 7,
      "minRange": 2400,
      "maxRange": 5800,
      "minMil": 1382.0,
      "maxMil": 800.0,
      "minTime": 62.9,
      "maxTime": 45.5
    },
    {
      "gunPower": 8,
      "minRange": 2400,
      "maxRange": 6350,
      "minMil": 1402.0,
      "maxMil": 800.0,
      "minTime": 68.6,
      "maxTime": 49.4
    },
  ];

  weapons.addAll(generateIntervalData("MA8", ma8Data, id!));
  id += ((ma8Data.length *
      ((ma8Data[0]["maxRange"] - ma8Data[0]["minRange"]) ~/ 50)) as int?)!;

  // Add 120MM data
  const List<Map<String, dynamic>> mm120Data = [
    {
      "gunPower": 1,
      "minRange": 600,
      "maxRange": 1850,
      "minMil": 1432,
      "maxMil": 800,
      "minTime": 27.0,
      "maxTime": 19.0
    },
    {
      "gunPower": 2,
      "minRange": 1000,
      "maxRange": 3000,
      "minMil": 1427,
      "maxMil": 800,
      "minTime": 36.0,
      "maxTime": 26.0
    },
    {
      "gunPower": 3,
      "minRange": 1400,
      "maxRange": 4100,
      "minMil": 1423,
      "maxMil": 800,
      "minTime": 43.0,
      "maxTime": 29.0
    },
    {
      "gunPower": 4,
      "minRange": 1900,
      "maxRange": 5250,
      "minMil": 1411,
      "maxMil": 800,
      "minTime": 51.0,
      "maxTime": 37.0
    },
    {
      "gunPower": 5,
      "minRange": 2300,
      "maxRange": 6100,
      "minMil": 1403,
      "maxMil": 808,
      "minTime": 56.2,
      "maxTime": 41.0
    },
    {
      "gunPower": 6,
      "minRange": 2600,
      "maxRange": 7000,
      "minMil": 1398,
      "maxMil": 800,
      "minTime": 61.0,
      "maxTime": 45.0
    },
    {
      "gunPower": 7,
      "minRange": 2900,
      "maxRange": 7650,
      "minMil": 1219,
      "maxMil": 800,
      "minTime": 69.0,
      "maxTime": 49.0
    },
    {
      "gunPower": 8,
      "minRange": 3200,
      "maxRange": 8150,
      "minMil": 1395,
      "maxMil": 800,
      "minTime": 74.0,
      "maxTime": 53.0
    },
    {
      "gunPower": 9,
      "minRange": 3200,
      "maxRange": 8500,
      "minMil": 1403,
      "maxMil": 800,
      "minTime": 80.0,
      "maxTime": 57.0
    },
  ];
  weapons.addAll(generateIntervalData("120MM", mm120Data, id));

  return weapons;
}

List<Weapon> generateIntervalData(
    String name, List<Map<String, dynamic>> data, int startId) {
  List<Weapon> intervalWeapons = [];
  int id = startId;

  for (var entry in data) {
    for (double range = entry["minRange"].toDouble();
        range <= entry["maxRange"].toDouble();
        range += 50.0) {
      double progress = (range - entry["minRange"].toDouble()) /
          (entry["maxRange"].toDouble() - entry["minRange"].toDouble());
      double mil = entry["minMil"].toDouble() +
          (entry["maxMil"].toDouble() - entry["minMil"].toDouble()) * progress;
      double flightTime = entry["minTime"].toDouble() +
          (entry["maxTime"].toDouble() - entry["minTime"].toDouble()) *
              progress;

      // Round values to two decimal places
      double roundedRange = double.parse(range.toStringAsFixed(2));
      double roundedMil = double.parse(mil.toStringAsFixed(2));
      double roundedFlightTime = double.parse(flightTime.toStringAsFixed(2));

      // Assign gunPower from the current data entry
      int gunPower = entry["gunPower"];

      intervalWeapons.add(Weapon(
        name: name,
        range: roundedRange,
        mil: roundedMil,
        flightTime: roundedFlightTime,
        gunPower: gunPower,
        id: id++,
      ));
    }
  }
  return intervalWeapons;
}

File: lib/views/map/weapon.py
import pandas as pd
import numpy as np

# Define the weapon data
weapon_data = [
    # MA7 Weapon Data
    {"weapon": "MA7", "charge": 0, "min_range": 250, "max_range": 750, "min_mil": 1427, "max_mil": 800, "min_flight_time": 18.3, "max_flight_time": 13.1},
    {"weapon": "MA7", "charge": 1, "min_range": 500, "max_range": 1550, "min_mil": 1437, "max_mil": 800, "min_flight_time": 25.3, "max_flight_time": 18.2},
    {"weapon": "MA7", "charge": 2, "min_range": 800, "max_range": 2150, "min_mil": 1408, "max_mil": 952, "min_flight_time": 25.9, "max_flight_time": 18.6},
    {"weapon": "MA7", "charge": 3, "min_range": 1000, "max_range": 2900, "min_mil": 1421, "max_mil": 800, "min_flight_time": 40.2, "max_flight_time": 28.8},
    {"weapon": "MA7", "charge": 4, "min_range": 1300, "max_range": 3480, "min_mil": 1405, "max_mil": 800, "min_flight_time": 46.1, "max_flight_time": 33.1},
    {"weapon": "MA7", "charge": 5, "min_range": 1600, "max_range": 4050, "min_mil": 1394, "max_mil": 800, "min_flight_time": 51.5, "max_flight_time": 37.2},
    {"weapon": "MA7", "charge": 6, "min_range": 1800, "max_range": 4600, "min_mil": 1395, "max_mil": 800, "min_flight_time": 57, "max_flight_time": 41},

    # MA8 Weapon Data
    {"weapon": "MA8", "charge": 0, "min_range": 200, "max_range": 550, "min_mil": 1410, "max_mil": 800, "min_flight_time": 14, "max_flight_time": 10},
    {"weapon": "MA8", "charge": 1, "min_range": 400, "max_range": 1125, "min_mil": 1414, "max_mil": 800, "min_flight_time": 21, "max_flight_time": 15.1},
    {"weapon": "MA8", "charge": 2, "min_range": 600, "max_range": 1650, "min_mil": 1410, "max_mil": 800, "min_flight_time": 27.2, "max_flight_time": 17.6},
    {"weapon": "MA8", "charge": 3, "min_range": 900, "max_range": 2700, "min_mil": 1427, "max_mil": 800, "min_flight_time": 36.4, "max_flight_time": 26.1},
    {"weapon": "MA8", "charge": 4, "min_range": 1300, "max_range": 3550, "min_mil": 1409, "max_mil": 800, "min_flight_time": 44.2, "max_flight_time": 45},
    {"weapon": "MA8", "charge": 5, "min_range": 1600, "max_range": 4400, "min_mil": 1412, "max_mil": 868, "min_flight_time": 51.1, "max_flight_time": 39.1},
    {"weapon": "MA8", "charge": 6, "min_range": 2100, "max_range": 5150, "min_mil": 1387, "max_mil": 800, "min_flight_time": 57.8, "max_flight_time": 41.8},
    {"weapon": "MA8", "charge": 7, "min_range": 2400, "max_range": 5800, "min_mil": 1382, "max_mil": 800, "min_flight_time": 62.9, "max_flight_time": 45.5},
    {"weapon": "MA8", "charge": 8, "min_range": 2400, "max_range": 6350, "min_mil": 1402, "max_mil": 800, "min_flight_time": 68.6, "max_flight_time": 49.4}
]

# Calculate interpolated values for every 50 meters
rows = []
for data in weapon_data:
    ranges = np.arange(data["min_range"], data["max_range"] + 50, 50)
    mils = np.interp(ranges, [data["min_range"], data["max_range"]], [data["min_mil"], data["max_mil"]])
    flight_times = np.interp(ranges, [data["min_range"], data["max_range"]], [data["min_flight_time"], data["max_flight_time"]])
    for r, m, t in zip(ranges, mils, flight_times):
        rows.append({
            "Weapon": data["weapon"],
            "Charge": data["charge"],
            "Range (m)": r,
            "Mil": round(m, 2),
            "Flight Time (s)": round(t, 2)
        })

# Create a DataFrame and save as CSV
df = pd.DataFrame(rows)
file_path = "/mnt/data/weapon_details_interpolated.csv"
df.to_csv(file_path, index=False)
file_path
